executor/src/witgen/jit/processor.rs

```diff
@@ -293,6 +320,78 @@ impl<'a, T: FieldElement, FixedEval: FixedEvaluator<T>> Processor<'a, T, FixedEv
     }
 }
 
+/// Returns all variables occurring in the identity.
+/// This also resolves intermediate polynomials.
+fn variables_in_identity<'a, T: FieldElement>(
+    fixed_data: &'a FixedData<'a, T>,
+    identity: &'a Identity<T>,
+    row_offset: i32,
+) -> impl Iterator<Item = Variable> + 'a {
+    identity
+        .children()
+        .flat_map(move |e| variables_in_expression(fixed_data, e, row_offset))
+}
+
+/// Returns all variables occurring in the expression.
+/// This also resolves intermediate polynomials.
+fn variables_in_expression<'a, T: FieldElement>(
+    fixed_data: &'a FixedData<'a, T>,
+    expression: &'a Expression<T>,
+    row_offset: i32,
+) -> impl Iterator<Item = Variable> + 'a {
+    expression
+        .all_children()
+        .flat_map(move |e| -> Box<dyn Iterator<Item = Variable> + 'a> {
+            match e {
+                Expression::Reference(r) => match r.poly_id.ptype {
+                    PolynomialType::Constant => Box::new(std::iter::empty()),
+                    PolynomialType::Committed => {
+                        Box::new(std::iter::once(Variable::from_reference(r, row_offset)))
+                    }
+                    PolynomialType::Intermediate => {
+                        let definition = &fixed_data.intermediate_definitions[&r.to_thin()];
+                        Box::new(variables_in_expression(fixed_data, definition, row_offset))
```

Comment:
This would take exponentially long for the folding constraint :/

You'd need to cache the result for a given intermediate definition. I tried to look into a way to abstract this away in #2312, as it always comes up when dealing with intermediate polynomials, but I was pretty unhappy with the verbosity...

Maybe it's also fine for now, because I don't think we'll generate bus witnesses with JIT (the handwritten witgen is actually quite robust), but at least we should be aware.